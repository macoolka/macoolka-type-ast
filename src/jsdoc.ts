import { parse, Annotation, Tag } from 'doctrine';
import { compareString, arrayCompare } from 'macoolka-compare'
import { Lens } from 'monocle-ts'
import * as A from 'fp-ts/lib/Array';
import { flow } from 'fp-ts/lib/function'
import { Option } from 'fp-ts/lib/Option'
import { pipe } from 'fp-ts/lib/pipeable'
import { array } from 'macoolka-collection'
export { Annotation }
const lensTitle = Lens.fromProp<Tag>()('title');

export const parseJSDoc = (source: string): Annotation => {
    return parse(source, { unwrap: true });
};

export const hasTag = (tag: string[]) => (annotation: Annotation): boolean =>

    arrayCompare().contains_every(annotation.tags.map(lensTitle.get))(tag)

export const getTag = (tag: string) => (annotation: Annotation): Option<Tag> =>
    pipe(
        annotation.tags,
        A.findFirst(flow(lensTitle.get, compareString.eq(tag)))
    )
export const tagToRecord = (a: Annotation) =>
    pipe(
        a.tags,
        array.toRecord('title')
    )

