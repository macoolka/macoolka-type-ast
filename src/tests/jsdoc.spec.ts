import {parseJSDoc,hasTag,getTag,tagToRecord} from '../jsdoc';
import * as O from 'fp-ts/lib/Option'
const a=`
 The property specifies the alignment for the selected item inside the flexible container.
 The property overrides the default alignment set by the container's align property.
`
const b=`
The property specifies the alignment for the selected item inside the flexible container.
The property overrides the default alignment set by the container's align property.
@typeclass
@data
`
const c=`
The property specifies the alignment for the selected item inside the flexible container.
The property overrides the default alignment set by the container's align property.
@typeclass
`
const d=`
The property specifies the alignment for the selected item inside the flexible container.
The property overrides the default alignment set by the container's align property.
@typeclass
@exmaple
abc
@since 0.1.1
@ZH
a

c
`
const pa=parseJSDoc(a);
const pb=parseJSDoc(b);
const pc=parseJSDoc(c);
const pd=parseJSDoc(d);
describe('jsdoc',()=>{
    it('parseJSDoc',()=>{
        expect(pa).toMatchSnapshot();
        expect(pb).toMatchSnapshot();
    })
    it('hasTag',()=>{
        expect(hasTag(['typeclass'])(pb)).toBeTruthy();
        expect(hasTag(['data'])(pb)).toBeTruthy();
        expect(hasTag(['data','typeclass'])(pb)).toBeTruthy();
        expect(hasTag(['typeclass'])(pc)).toBeTruthy();
        expect(hasTag(['data'])(pc)).toBeFalsy();
        expect(hasTag(['data','typeclass'])(pc)).toBeFalsy();
        expect(hasTag(['dta'])(pb)).toBeFalsy();
        expect(hasTag(['data'])(pa)).toBeFalsy();
    })  
    it('getTag',()=>{
        expect(O.isSome(getTag('typeclass')(pb))).toBeTruthy();
    }) 
    it('tagToRecord',()=>{
       
        expect(tagToRecord(pd)).toMatchSnapshot();
    })
})