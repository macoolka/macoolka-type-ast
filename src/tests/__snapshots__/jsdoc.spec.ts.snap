// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsdoc parseJSDoc 1`] = `
Object {
  "description": "The property specifies the alignment for the selected item inside the flexible container.
The property overrides the default alignment set by the container's align property.",
  "tags": Array [],
}
`;

exports[`jsdoc parseJSDoc 2`] = `
Object {
  "description": "The property specifies the alignment for the selected item inside the flexible container.
The property overrides the default alignment set by the container's align property.",
  "tags": Array [
    Object {
      "description": null,
      "title": "typeclass",
    },
    Object {
      "description": null,
      "title": "data",
    },
  ],
}
`;

exports[`jsdoc tagToRecord 1`] = `
Object {
  "ZH": Object {
    "description": "a

c",
  },
  "exmaple": Object {
    "description": "abc",
  },
  "since": Object {
    "description": "0.1.1",
  },
  "typeclass": Object {
    "description": null,
  },
}
`;
